### DISCUSSION TITLE
How to account for version managers?

### DISCUSSION THREAD

**Discussion by lindboe** [johnlindquist/kit#1273] [Q&A]
I have a script where I want to run a Ruby gem-based command (`pod install`) inside a project. This works inside the project, because I'm using rbenv, and I've already installed gems for the version in use there.

However, scriptkit is not aware of rbenv, and fails because gems are not installed for the system version of Ruby.

I would love for scriptkit to be able to use the version managers I've already configured in my zsh environment. (This will also eventually apply for nvm). Does anyone have recommendations on how to manage this?

I should note: in order to enable sharing of this script, I would _love_ to have a solution that is version-manager-agnostic. That is, a solution that does not hardcode a path to a particular tool. So, ideally, I'd love the ability to execute this script in my usual shell, not sure if there's a tool meant for that in scriptkit.

**Comment by johnlindquist** [comment]
Could you use something like:

await exec('rbenv exec ruby your_script.rb')

I'm not familiar with loading in rbenv, but I just assume there's something you can add to your script to grab the environment

**Reply by lindboe** [reply]
Not with rbenv, because it requires a shim script to run so that it's able to locate the correct versions of gems. (Without that, I think you have to run `rbenv rehash` after the first time you install a gem; not sure if that actually works, it's not the typical flow).

This script is typically configured to run by adding configuration to your shell's bashrc equivalent (e.g., `eval "$(~/.rbenv/bin/rbenv init - zsh)` for zsh). If there's an init script for kit's shell, I could do this, but it's less than ideal; this really interferes with my ability to share this script because developers in this ecosystem (React Native) haven't standardized on using a ruby version manager.

### The rest of my exploration

I just went on a bit of a journey trying to dig into this more.

For context on the script itself: I'm trying to build a script for common debugging flows for React Native builds. There are scripts for building it the usual way, but not for the manual steps I frequently take when debugging a build issue. So I was hoping to make a shareable script where users can check off which debugging steps they want to run, and remember those for next time, after they make a change to test.

To summarize my exploration of this problem:

1. I was getting confused about which various things are using what underlying systems. I had thought `exec` was using `zx`, but it looks like it's actually the library `execa`. But I had been trying to use `zx`'s `$.shell` and it wasn't doing anything 😄 
2. I eventually figured out I can use `exec(..., {shell: '/bin/zsh'}`. It's a start! I could get the value of `$SHELL` from another exec and pass it in, so whatever shell the user is typically using when building a React Native project can be used.
3. During this time, a realization: `.zshrc` is only for interactive shells. I tried using `.zshenv` instead, and this _does_ get loaded with `exec` and the `shell` option.
4. However: my version of `rbenv` is installed with homebrew. homebrew config is usually run in `.zprofile`, which is _not_ run in this non-interactive shell. I could make it be by adding `source ~/.zprofile`, but that makes this even less portable.
5. To achieve my goal of portability, I decided to try `terminal(...)` to see if that can just let each user have the terminal setup they're used to and already works. This found and started running the command successfully. But in a new twist, there's more expected shell config that cocoapods won't work without:

```
    [33mWARNING: CocoaPods requires your terminal to be using UTF-8 encoding.
    Consider adding the following to ~/.profile:

    export LANG=en_US.UTF-8
    bundler: failed to load command: pod (/Users/lizzi/dev/ExpensifyApp/vendor/bundle/ruby/2.7.0/bin/pod)
Encoding::CompatibilityError: Unicode Normalization not appropriate for ASCII-8BIT
  /Users/lizzi/.rbenv/versions/2.7.5/lib/ruby/2.7.0/unicode_normalize/normalize.rb:141:in `normalize'
```

I haven't had a chance yet to look into how that should be configured for this terminal, and at this point I'm thinking that it makes more sense for me to just write a script the user should run from their usual terminal, where everything is configured to work. Which I'm bummed about, because I was excited to use kit's database features to save "most recent" configuration choices easily.

**Reply by lindboe** [reply]
Mmm, yup, I added `env: {"LANG": "en_US.UTF-8"}` to the `exec` options and now I hit another `$PATH`-related issue, I assume there's some additional path config I need in my `.zshrc`. Just dealing with a lot of tools that I've never needed to run in a non-interactive shell, and their setup instructions did not account for that use case.

**Reply by johnlindquist** [reply]
@lindboe 

My assumption was waaaay off. I apologize for trivializing something I didn't understand.

Kit does have a Built-in interactive terminal:

```
await term()
```

But it sounds like you want this to run in the background... `await term()` uses the "node-pty" library under the hood (It's the same library vs code, hyper, etc use).

Depending on the goal of your script, "node-pty" might be an option since it's an interactive shell you can run from a script. Handling the output gets _reeeeeeaaaally_ complicated since it will send all the ansi codes for colors/cursor movement/etc, so it's not a very reliable solution if you care about parsing the output unless you're just looking for a specific string.

```js
// Name: node-pty demo

import "@johnlindquist/kit"
import * as pty from "node-pty"
import stripAnsi from "strip-ansi"

// You need to manually install/compile with the built-in node
// Since node-pty is a native module, it needs to be compiled with same node version the Kit.app is using:
/*
cd ~/.kenv && PATH=~/.knode/bin:$PATH npm i -D node-pty
*/

let pseudoTerm = pty.spawn(process.env.SHELL, [], {})

let filePath = tmpPath("test.txt")

let endOnSettle = _.debounce(() => {
  edit(filePath)
  pseudoTerm.kill()
}, 1000)

pseudoTerm.onData(async data => {
  // convert to string and _attempt_ to remove non-printable characters
  let str = stripAnsi(data.toString())
  appendFile(filePath, str)

  // end term if we haven't received any data in 1 second
  endOnSettle()
})

pseudoTerm.write(`ls -la\n`) // you need \n to send the command
```

**Reply by lindboe** [reply]
Oh that's super cool! In this particular case, the output is ignorable if it works, but if it fails it will be important to show the logs to the user. For this reason I was thinking about sticking with foregrounding it.

For context, the reason I was starting with `exec` instead of `terminal` was because I saw the recommendation floating around to only use `terminal` if the action requires user interaction with the terminal. From a UX perspective, this doesn't, but it does seem to basically require an interactive shell.
